### ISSUE TITLE
Explicitly set cache instruction key

### ISSUE THREAD

**Issue by fkorotkov** [cirruslabs/cirrus-ci-docs#867]
# Description

Right now the only way to set a cache key is via `fingerprint_script` which output will `SHA256`ed and the result will be used as the cache key: `<CACHE_NAME>_<SHA256(fingerprint_script ouput)>`.

For using with [Starlark](https://cirrus-ci.org/guide/programming-tasks/) it might be useful to give full control over the key to users and add a `key` filed to cache instructions.

Let's also expose a GraphQL query for retrieving cache info for a particular repository and a cache key.

# Context

With this feature and https://github.com/cirruslabs/cirrus-cli/issues/398 we'll be able to workaround a use case discovered in https://github.com/ponylang/ponyc/pull/3772 where cache population requires a lot of resources but happens very rarely.

With these features one will be able to write something:

```python
load("github.com/cirrus-modules/graphql", "cache_info")
load("cirrus", "fs", "hash")

_cache_buster = "2021-06-07"

def main():
  cache_key = hash.md5(
    _cache_buster + fs.read("lib/CMakeLists.txt")
  )
  env = {
    "LLVM_CACHE_KEY": cache_key
    "CPU": 2
    "MEMORY": "4G"
  }
  if cache_info(cache_key) is None:
    env["CPU"] = 8
    env["MEMORY"] = "24G"
  return {
    "env": env
  }
```

Then in the `.cirrus.yml` it will be possible to have something like that:

```yaml
libs_cache:    
  folder: build/libs    
  key:  "$LLVM_CACHE_KEY" 
  populate_script: make libs build_flags=-j8
```

**Comment by fkorotkov** [comment]
`.star` is evaluated before parsing of the `.yml` config. Essentially `.star` is evaluated and then appended to `.yml` before passing it to the config parser which will output tasks.

So it won't start a new VM, it will figure out the right size even before starting a VM.

**Comment by SeanTAllen** [comment]
Cool. That would indeed fit what we need and had been wishing existed.

**Comment by SeanTAllen** [comment]
@fkorotkov does this being closed and the associated PR mean we can update our CI usage for the ponyc repo?

**Comment by fkorotkov** [comment]
Yes, you can. The change was deployed on Friday. FYI there are also greedy containers now which can help with extra CPU for the cache population https://medium.com/cirruslabs/introducing-greedy-container-instances-29aad06dc2b4

**Comment by SeanTAllen** [comment]
@fkorotkov it seems 

`load("github.com/cirrus-modules/graphql", "cache_info")`

isn't valid for the graphql module.

what's the proper way to do the cache check?

**Comment by fkorotkov** [comment]
My bad. I mislead you. We only implemented the explicit cache key and not the `cache_info`. There is an isssue with `cache_info` implementation that we need to know ahead of time where a corresponding task will be executed: whether it's on infrastructure managed by us or if the execution will be on a connected cloud. This information is required to get the right credentials and look in the proper blog storage like GCS, S3, etc.

We didn't find a way elegantly solve this problem so we decided to not do it until there is a better option.

For now, I'd encourage you to try the greedy tasks and lower CPU requirements so in case of a cache miss the task will most likely be able to use more CPU for cache population.
